### PR TITLE
containerapp refactor decorator

### PR DESCRIPTION
**Pull request by Greedygre**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

### Related command

az containerapp create

### General Guidelines

- [ ] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [ ] Have you run `python scripts/ci/test_index.py -q` locally?

For new extensions:

- [ ] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).

### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your pull request is merged into main branch, a new pull request will be created to update `src/index.json` automatically.  
You only need to update the version information in file setup.py and historical information in file HISTORY.rst in your PR but do not modify `src/index.json`.

### CODE REVIEW

**Comment by yonzhan** [comment]
Thank you for your contribution! We will review the pull request and get back to you soon.

File: src/containerapp-preview/azext_containerapp_preview/containerapp_decorator.py [hunk 1]
```diff
@@ -40,29 +40,21 @@ def __init__(
...
+        super().construct_containerapp()
+        self.set_up_extended_location()
 
-    def set_up_extended_location(self, containerapp_def):
+    def set_up_extended_location(self):
         if self.get_argument_environment_type() == CONNECTED_ENVIRONMENT_TYPE:
             parsed_env = parse_resource_id(self.get_argument_managed_env())  # custom_location check here perhaps
             env_name = parsed_env['name']
             env_rg = parsed_env['resource_group']
             env_info = self.get_environment_client().show(cmd=self.cmd, resource_group_name=env_rg, name=env_name)
```
Comment (Juliehzl) [inline src/containerapp-preview/azext_containerapp_preview/containerapp_decorator.py:51]: only GET when there is no extended location

File: src/containerapp/azext_containerapp/containerapp_decorator.py [hunk 2]
```diff
@@ -506,58 +507,56 @@ def construct_containerapp(self):
...
             template_def["terminationGracePeriodSeconds"] = self.get_argument_termination_grace_period()
 
-        containerapp_def = ContainerAppModel
-        containerapp_def["location"] = location
-        containerapp_def["identity"] = identity_def
-        containerapp_def["properties"]["environmentId"] = self.get_argument_managed_env()
-        containerapp_def["properties"]["configuration"] = config_def
-        containerapp_def["properties"]["template"] = template_def
-        containerapp_def["tags"] = self.get_argument_tags()
+        self.containerapp_def = ContainerAppModel
```
Comment (Juliehzl) [inline src/containerapp/azext_containerapp/containerapp_decorator.py:510]: remove this line

File: src/containerapp/azext_containerapp/containerapp_decorator.py [hunk 3]
```diff
@@ -506,58 +507,56 @@ def construct_containerapp(self):
...
                 no_wait=self.get_argument_no_wait())
 
             return r
         except Exception as e:
             handle_raw_exception(e)
 
-    def construct_containerapp_for_post_process(self, containerapp_def, r):
+    def construct_containerapp_for_post_process(self, r):
         if is_registry_msi_system(self.get_argument_registry_identity()):
             while r["properties"]["provisioningState"] == "InProgress":
```
Comment (Juliehzl) [inline src/containerapp/azext_containerapp/containerapp_decorator.py:541]: consider using polling

**Review by Juliehzl — APPROVED** [review]

File: src/containerapp/azext_containerapp/containerapp_decorator.py (hunk 3 above)
Comment (Greedygre) [reply src/containerapp/azext_containerapp/containerapp_decorator.py:541]: consider in next PR

**Comment by Greedygre** [comment]
Hi @wangzelin007 
I found when the containerapp and containerapp-preview run together, they affect each other.
Because when containerapp test finished, it will remove containerapp extension. I re-tried twice, and check failed randomly.
Counld you help to handle it?
```
Test on modules: containerapp
Test on modules: containerapp-preview
```

**Comment by Greedygre** [comment]
Hi @zhoxing-ms 
Zunli has help to review and approved this PR, could you help to review and merge? Thanks!

**Comment by wangzelin007** [comment]
> Hi @wangzelin007 I found when the containerapp and containerapp-preview run together, they affect each other. Because when containerapp test finished, it will remove containerapp extension. I re-tried twice, and check failed randomly. Could you help to handle it? Thanks.
> 
> ```
> Test on modules: containerapp
> Test on modules: containerapp-preview
> ```
> 
> image

The subprocess package will create a subprocess to execute commands, which causes the test to start before the installation is complete, so I added a 2min waiting time.
This problem also only occurs in containerapp, because other extensions are installed by `az extension add`.

**Review by zhoxing-ms — APPROVED** [review]

**Pull request merged**